### PR TITLE
rename access-key to admin-secret (close #1347)

### LINKED ISSUE #1347: X-Hasura-Access-Key should be renamed to X-Hasura-Secret-Key to make it more obvious
Maybe X-Hasura-Access-Key should be renamed to X-Hasura-Secret-Key to make it more obvious. Like all oauth providers (google, github etc) tend to call their private keys secret_<provider>_key and their public key public_<provider>_id.

Also, maybe the permissions tab should be disabled when access key is not set to make it more obvious.

### PR DESCRIPTION
**Pull request by nizar-m**
### Description
Rename the admin secret key header used to access GraphQL engine from X-Hasura-Access-Key to X-Hasura-Admin-Secret

What component does this PR affect? 

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [x] Console
- [x] CLI
- [x] Docs
- [x] Community Content
- [ ] Build System

### Related Issue

#1347 

### Solution and Design

- [ ] Server
  -  [x] Change envVar HASURA_GRAPHQL_ACCESS_KEY to HASURA_GRAPHQL_ADMIN_SECRET
   - [x] Change flag --access-key to --admin-secret
   - [x] Change the admin secret key header from X-Hasura-Access-Key to X-Hasura-Admin-Secret
   - [x] Server will still support the following (even though they are deprecated)
       - [x] header X-Hasura-Access-Key
       - [x] EnvVar HASURA_GRAPHQL_ACCESS_KEY
       - [x] flag --access-key
- [x] Update references in the docs
   - [ ] Change pngs `dev-mode-role-header-access-key.png` and `access-key-console.png`
- [ ] Console
- [x] Cli
   - [x] Change cli flags --access-key to --admin-secret
   - [x] Tests with hasura cli for both --access-key and --admin-secret flags
   - [x] Test config.yaml with both access_key and admin_secret
- [ ] Install manifests
   - [ ] Azure
   - [ ] Docker run
   - [ ] Docker compose
   - [ ] Kubernetes 
   - [ ] DigitalOcean
### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] Community content

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [x] This change requires a change in the documentation. 
- [x] I have updated the documentation accordingly.
- [ ] I have added required tests.

### CODE REVIEW

**Comment by shahidhk** [comment]
Reviews required from @0x777 @shahidhk and @praveenweb before merging.

**Comment by praveenweb** [comment]
@nizar-m - console has to be backwards compatible. Global replace of `x-hasura-access-key` to `x-hasura-admin-secret` might not work.

File: console/src/Globals.js [hunk 1]
```diff
@@ -25,8 +25,8 @@ const globals = {
   dataApiUrl: checkExtraSlashes(window.__env.dataApiUrl),
   devDataApiUrl: window.__env.devDataApiUrl,
   nodeEnv: window.__env.nodeEnv,
-  accessKey: window.__env.accessKey,
-  isAccessKeySet: window.__env.isAccessKeySet,
+  adminSecret: window.__env.adminSecret,
```
Comment (praveenweb) [inline console/src/Globals.js]: cli sends this adminSecret env. But console needs to check for accessKey env for older versions of cli.

File: console/src/Globals.js [hunk 2]
```diff
@@ -46,12 +46,12 @@ if (!window.__env.consoleMode) {
...
-if (!window.__env.accessKey) {
-  globals.accessKey = null;
+if (!window.__env.adminSecret) {
+  globals.adminSecret = null;
 }
 
-if (!window.__env.isAccessKeySet) {
-  globals.isAccessKeySet = false;
+if (!window.__env.isAdminSecretSet) {
+  globals.isAdminSecretSet = false;
```
Comment (praveenweb) [inline console/src/Globals.js]: server sends this. But for older versions of server, check for isAccessKeySet

File: console/src/components/ApiExplorer/ApiRequest.js [hunk 3]
```diff
@@ -323,21 +320,21 @@ class ApiRequest extends Component {
               onBlur={this.handleBlur}
               data-test={`header-value-${i}`}
               type={
-                header.key.toLowerCase() === 'x-hasura-access-key' &&
-                !this.state.accessKeyVisible
+                header.key.toLowerCase() === 'x-hasura-admin-secret' &&
```
Comment (praveenweb) [inline console/src/components/ApiExplorer/ApiRequest.js]: check for both x-hasura-admin-secret and x-hasura-access-key

File: console/src/components/Common/validateLogin.js [hunk 4]
```diff
@@ -9,13 +9,13 @@ import { changeRequestHeader } from '../ApiExplorer/Actions';
...
 
-const checkValidity = accessKey => {
+const checkValidity = adminSecret => {
   return dispatch => {
     const url = Endpoints.getSchema;
     const currentSchema = 'public';
     const headers = {
       'Content-Type': 'application/json',
-      'X-Hasura-Access-Key': accessKey,
+      'X-Hasura-Admin-Secret': adminSecret,
```
Comment (praveenweb) [inline console/src/components/Common/validateLogin.js]: Also send another header `x-hasura-access-key` for backwards compatibility.

File: console/src/components/Common/validateLogin.js [hunk 5]
```diff
@@ -40,29 +40,29 @@ const checkValidity = accessKey => {
 
 const validateLogin = ({ dispatch }) => {
```
Comment (praveenweb) [inline console/src/components/Common/validateLogin.js]: accept another argument for version and do version check below.

File: console/src/components/Login/Login.js [hunk 6]
```diff
@@ -52,11 +52,11 @@ class Login extends Component {
                 >
                   <div className={'input-group ' + styles.input_group}>
                     <input
-                      onChange={this.handleAccessKey}
+                      onChange={this.handleAdminSecret}
                       className={styles.form_input + ' form-control'}
                       placeholder="Password"
                       type="password"
-                      placeholder="Enter access key"
+                      placeholder="Enter admin secret"
```
Comment (praveenweb) [inline console/src/components/Login/Login.js]: Placeholder text based on server version

File: console/src/Globals.js [hunk 7]
```diff
@@ -25,8 +25,14 @@ const globals = {
   dataApiUrl: checkExtraSlashes(window.__env.dataApiUrl),
   devDataApiUrl: window.__env.devDataApiUrl,
   nodeEnv: window.__env.nodeEnv,
-  accessKey: window.__env.accessKey,
-  isAccessKeySet: window.__env.isAccessKeySet,
+  adminSecret: window.__env.adminSecret || window.__env.accessKey,
+  isAdminSecretSet:
+    window.__env.isAdminSecretSet || window.__env.isAccessKeySet,
+  adminSecretName:
```
Comment (praveenweb) [inline console/src/Globals.js]: change this to adminSecretLabel?

File: console/src/components/ApiExplorer/ApiRequest.js [hunk 8]
```diff
@@ -323,23 +321,25 @@ class ApiRequest extends Component {
...
+              {header.key.toLowerCase() ===
+              `x-hasura-${globals.adminSecretName}` ? (
                 <i
-                  className={styles.showAccessKey + ' fa fa-eye'}
-                  data-header-id={i}
-                  aria-hidden="true"
-                  onClick={this.onShowAccessKeyClicked.bind(this)}
-                />
-              ) : null}
+                    className={styles.showAccessKey + ' fa fa-eye'}
```
Comment (praveenweb) [inline console/src/components/ApiExplorer/ApiRequest.js]: change className here to `showAdminSecret` and apply the same to scss file?

**Comment by shahidhk** [comment]
CLI needs more changes to handle older versions.

**Comment by coco98** [comment]
How are we testing backwards compatibility for this? One idea is that tests that we have that use access key can be duplicated with admin-secret?
@rakeshkky @shahidhk @wawhal ?

**Comment by shahidhk** [comment]
@coco98 Yes, server and cli has tests that use both `access-key` and `admin-secret` and you can see it is failing right now since backward compatibility for CLI is still being worked on.

**Comment by wawhal** [comment]
@coco98 we can't run console tests for both `access-key` and `admin-secret`. The build time will go up big time. We have checked the backwards compatibility in #1559 where tests are run with `access-key`.

**Comment by shahidhk** [comment]
@praveenweb @wawhal When opening console from CLI (with admin secret set), the console is making a request to the server with `x-hasura-admin-secret` header irrespective of the server version. This will show login prompt on older version.

Console should check for server version and use the right header accordingly.

**Comment by shahidhk** [comment]
@nizar-m Are you adding `test-deprecated-server-flags.sh`?

**Comment by dsandip** [comment]
@shahidhk @praveenweb Are CLI/Console reviews needed? If yes, please do so?

File: console/src/helpers/Html.js [hunk 9]
```diff
@@ -43,8 +43,9 @@ export default class Html extends Component {
                 apiPort: '${process.env.API_PORT}',
                 dataApiUrl: '${process.env.DATA_API_URL}',
                 devDataApiUrl: '${process.env.DEV_DATA_API_URL}',
-                accessKey: '${process.env.ACCESS_KEY}',
-                isAccessKeySet: ${process.env.IS_ACCESS_KEY_SET},
+                adminSecret: '${process.env.ACCESS_KEY}',
+                isAdminKeySet: ${process.env.IS_ADMIN_SECRET_ET ||
```
Comment (praveenweb) [inline console/src/helpers/Html.js]: Typo. `process.env.IS_ADMIN_SECRET_ET` -> `process.env.IS_ADMIN_SECRET_SET`

**Comment by rizkyramadhan** [comment]
Does this means we should add X-Hasura-Admin-Secret to our graphql queries on each client ?

**Comment by shahidhk** [comment]
@rizkyramadhan You should not use AccessKey/AdminSecret to query data from client side. It is your super-admin-master-password. Anyone can wipe out your database if they have access to this key/secret.

Setup authentication/access control: https://docs.hasura.io/1.0/graphql/manual/auth/index.html

**Pull request merged**